src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java

```diff
@@ -304,6 +308,76 @@ private ConfiguredTarget createRule(
                     prerequisiteMap.values()))
             .build();
 
+    if (!"toolchain".equals(rule.getRuleClass())) {
```

Comment:
Similar to my other comment, I wonder if all this logic can be encapsulated in a new method in ConstraintSemantics (I guess at this point we're not concerned about top-level specifically)?

Reply:
It's not top-level, true. I'll look for a better place to put this.

Reply:
Given that `ConstraintSemantics` is an `interface` and that this new method is fairly self contained regardless of implementation, I'm leaning towards a static function on the interface. However, that introduces a circular dependency involving `analysis_cluster`. 

Making it a regular member function should theoretically work too, but requires some more digging that I haven't done yet. E.g. I think importing `RuleContext` in `ConstraintSemantics` creates another circular dependency involving `analysis_cluster`.

For now I moved this to a member function in `ConfiguredTargetFactory`.